### ISSUE TITLE
`qml.Hamiltonian` mishandles operators multiplied by `qml.Identity`

### ISSUE THREAD

**Issue by Qottmann** [PennyLaneAI/pennylane#4382]
Expected behavior:

```python

ops = [qml.PauliX(0), qml.PauliX(1), qml.PauliX(0) @ qml.PauliX(1)]
H0 = qml.dot(np.ones(3), ops)

H1 = qml.dot(np.ones(3), [qml.Identity(0) @ op for op in ops])

>>> qml.matrix(H0) -  qml.matrix(H1)
array([[0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.],
       [0., 0., 0., 0.]])
```

Actual behavior:

```python
ops = [qml.PauliX(0), qml.PauliX(1), qml.PauliX(0) @ qml.PauliX(1)]
H0 = qml.Hamiltonian(np.ones(3), ops)

H1 = qml.Hamiltonian(np.ones(3), [qml.Identity(0) @ op for op in ops])

>>> qml.matrix(H0) -  qml.matrix(H1)
array([[-1.+0.j,  0.+0.j,  1.+0.j,  0.+0.j],
       [ 0.+0.j, -1.+0.j,  0.+0.j,  1.+0.j],
       [ 1.+0.j,  0.+0.j, -1.+0.j,  0.+0.j],
       [ 0.+0.j,  1.+0.j,  0.+0.j, -1.+0.j]])
```
Seems to be specfic to `qml.Hamiltonian`

Bonus question: Is it documented somewhere when best to use `qml.Hamiltonian` and when to use `qml.dot`?

**Comment by timmysilv** [comment]
I'm still not 100% sure what's causing this, but I will point out two quick things. First, `qml.dot` returns a `Sum`, not a `Hamiltonian`, so I guess it's bug-free! Second, matrix-multiplying a Tensor by another operator changes the _original tensor_, which is kinda unexpected. A quick example:
```pycon
>>> op = qml.PauliX(0) @ qml.PauliY(0)
>>> op @ qml.PauliZ(0), op
(PauliX(wires=[0]) @ PauliY(wires=[0]) @ PauliZ(wires=[0]),
 PauliX(wires=[0]) @ PauliY(wires=[0]) @ PauliZ(wires=[0]))  # the PauliZ is now part of op!
```

And in your case:
```pycon
>>> ops = [qml.PauliX(0), qml.PauliX(1), qml.PauliX(0) @ qml.PauliX(1)]
>>> H0 = qml.Hamiltonian(np.ones(3), ops)
>>> H0.ops, qml.matrix(H0)
([PauliX(wires=[0]), PauliX(wires=[1]), PauliX(wires=[0]) @ PauliX(wires=[1])],
 array([[0.+0.j, 1.+0.j, 1.+0.j, 1.+0.j],
        [1.+0.j, 0.+0.j, 1.+0.j, 1.+0.j],
        [1.+0.j, 1.+0.j, 0.+0.j, 1.+0.j],
        [1.+0.j, 1.+0.j, 1.+0.j, 0.+0.j]]))

>>> _ = [qml.Identity(0) @ op for op in ops]  # changes the last one in place only because the others aren't Tensors
>>> H0.ops, qml.matrix(H0)
([PauliX(wires=[0]),
  PauliX(wires=[1]),
  Identity(wires=[0]) @ PauliX(wires=[0]) @ PauliX(wires=[1])],
 array([[0.+0.j, 2.+0.j, 1.+0.j, 0.+0.j],
        [2.+0.j, 0.+0.j, 0.+0.j, 1.+0.j],
        [1.+0.j, 0.+0.j, 0.+0.j, 2.+0.j],
        [0.+0.j, 1.+0.j, 2.+0.j, 0.+0.j]]))
```

but that still shouldn't change the matrix so... idk

**Comment by timmysilv** [comment]
the assumption occurs [here](https://github.com/PennyLaneAI/pennylane/blob/bb74af81eaf1ff8e68701de108b19adb4cae8aa1/pennylane/ops/qubit/hamiltonian.py#L421-L424) btw. If a single term has overlapping wires, it'll just take the first sub-term (?) on that wire.

**Comment by timmysilv** [comment]
gonna close this because the `Tensor` warning is raised when this funky behaviour happens.

tl;dr if you have old-school `Tensor` operators with overlapping wires in a `Hamiltonian`, its matrix might be wrong. Consider fixing those tensor ops, or using the new-school `Prod` if necessary
